src/components/QueryEditor.tsx

```diff
@@ -1,56 +1,100 @@
-import { InlineField, Input, Select } from '@grafana/ui';
+import React, { ChangeEvent, KeyboardEvent, useEffect, useRef, useState } from 'react';
+import { InlineField, Select, TextArea } from '@grafana/ui';
 import { QueryEditorProps, SelectableValue } from '@grafana/data';
 import { CriblDataSource } from '../datasource';
 import { CriblDataSourceOptions, CriblQuery } from '../types';
 import { debounce } from 'lodash';
 
 type Props = QueryEditorProps<CriblDataSource, CriblQuery, CriblDataSourceOptions>;
 
-export function QueryEditor({ datasource, query, onChange, onRunQuery }: Props) {
-  const { maxResults } = query;
+const QUERY_TYPE_OPTIONS = ['saved', 'adhoc'].map((value) => ({ label: value, value }));
+const DEFAULT_QUERY_TYPE = 'adhoc';
 
+export function QueryEditor({ datasource, query, onChange, onRunQuery }: Props) {
   const debouncedOnRunQuery = useRef(debounce(onRunQuery, 750)).current;
   const [savedSearchIdOptions, setSavedSearchIdOptions] = useState<SelectableValue[]>([]);
+  const [savedSearchId, setSavedSearchId] = useState('');
 
-  const onSavedQueryIdChange = (sv: SelectableValue<string>) => {
-    const v = sv.value?.replace(/\s+/g, '') ?? ''; // auto-trim/remove any whitespace
-    if (v.match(/^[a-zA-Z0-9_]+$/)) {
-      onChange({ ...query, savedSearchId: v });
-      debouncedOnRunQuery();
+  const [queryType, setQueryType] = useState(query.type as string ?? DEFAULT_QUERY_TYPE);
+  const [adhocQuery, setAdhocQuery] = useState(query.type === 'adhoc' ? query.query : '');
+
+  const onQueryTypeChange = (sv: SelectableValue<string>) => {
+    const newQueryType = sv.value ?? DEFAULT_QUERY_TYPE;
+    setQueryType(newQueryType);
+    if (newQueryType === 'saved') {
+      onChange({ ...query, type: 'saved', savedSearchId });
+    } else {
+      onChange({ ...query, type: 'adhoc', query: adhocQuery });
+    }
+  };
+
+  const onAdhocQueryChange = (event: ChangeEvent<HTMLTextAreaElement>) => {
+    const newQuery = event.target.value;
+    setAdhocQuery(newQuery);
+    onChange({ ...query, type: 'adhoc', query: newQuery });
+    // Don't run it automatically, the user can hit Enter or click "Run query" when ready
+  };
+
+  const onAdhocQueryKeyDown = (event: KeyboardEvent) => {
+    if (event.key === 'Enter' && !event.shiftKey) { // allow shift-enter to add a line break
+      event.preventDefault();
+      if (adhocQuery.trim().length > 0) {
+        onRunQuery();
+      }
     }
   };
 
-  const onMaxResultsChange = (event: ChangeEvent<HTMLInputElement>) => {
-    onChange({ ...query, maxResults: Math.max(1, +event.target.value) });
-    debouncedOnRunQuery();
+  const onSavedQueryIdChange = (sv: SelectableValue<string>) => {
+    const newSavedSearchId = sv.value?.replace(/\s+/g, '') ?? ''; // auto-trim/remove any whitespace
+    setSavedSearchId(newSavedSearchId);
+    if (newSavedSearchId !== savedSearchId && newSavedSearchId.match(/^[a-zA-Z0-9_]+$/)) {
```

Comment:
Before running `setSavedSearchId` above.  Might be better to bail early

```suggestion
    if (newSavedSearchId === savedSearchId) return;
```

Reply:
This is deliberate, to "blank out" any previously selected id.  That way if the user clicks "Run" then it won't run an unexpected search.